course02/homework01/danielellis/linkedlist.h

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <memory>
+
+struct LinkedListNode
+{
+	LinkedListNode() = default;
+	LinkedListNode(const LinkedListNode& lln);
+	LinkedListNode(int payload);
+	std::shared_ptr<LinkedListNode> Clone(std::shared_ptr<LinkedListNode> last);
+	int& Retrieve(int indirectionsLeft);
+
+	std::shared_ptr<LinkedListNode> mNext;
+	int mPayload;
+};
+
+class LinkedList
+{
+public:
+	LinkedList();
+	LinkedList(const LinkedList& ll);
+	int Size() const;
+	int& At(int position);
+	int At(int position) const;
```

Comment:
Those methods would typically not be implemented on a singly linked list, as they cannot offer a constant complexity.